### PR TITLE
[SPARK-48714][PYTHON] Implement `DataFrame.mergeInto` in PySpark

### PR DESCRIPTION
**Pull request by xupefei**
### What changes were proposed in this pull request?

Spark 4.0 added a new `df.mergeInto` API, but it is missing from PySpark. This PR fixes that.
The support for this API in Spark Connect Python API will be added later by https://github.com/apache/spark/pull/46960.

### Why are the changes needed?

Because PySpark does not support `df.mergeInto`.

### Does this PR introduce _any_ user-facing change?

Yes, the user would be able to use the `df.mergeInto` API.

### How was this patch tested?

New unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@huaxingao @viirya would you mind taking a look please?

File: python/pyspark/sql/tests/test_dataframe.py [hunk 1]
```diff
@@ -741,6 +741,32 @@ def test_df_show(self):
...
+                source.mergeInto("testcat.ns1.target", source.id == col("target.id")) \
+                    .whenMatched(source.id == 1).update({"name": source.name}) \
+                    .whenNotMatched().insertAll() \
+                    .whenNotMatchedBySource().delete() \
+                    .merge()
+                self.assertEqual(
+                    self.spark.table("testcat.ns1.target").orderBy("id").collect(),
+                    [Row(id=1, name="Charlie"), Row(id=3, name="David")])
+        finally:
+            self.spark.conf.unset("spark.sql.catalog.testcat")
```
Comment (viirya) [inline python/pyspark/sql/tests/test_dataframe.py:767]: Maybe add more tests against all supported APIs under `whenMatched`, `whenNotMatched` and `whenNotMatchedBySource`?

**Review by viirya — COMMENTED** [review]
I quickly read through the implementation. Looks good overall. Maybe @huaxingao can take a look.

File: python/pyspark/sql/tests/connect/test_parity_dataframe.py [hunk 2]
```diff
@@ -54,6 +54,10 @@ def test_cached_property_is_copied(self):
     def test_toDF_with_schema_string(self):
         super().test_toDF_with_schema_string()
 
+    @unittest.skip("Spark Connect does not support mergeInto at the moment.")
```
Comment (allisonwang-db) [inline python/pyspark/sql/tests/connect/test_parity_dataframe.py:57]: Are we planning to support Spark Connect?

Comment (viirya) [reply python/pyspark/sql/tests/connect/test_parity_dataframe.py:57]: > The support for this API in Spark Connect Python API will be added later by https://github.com/apache/spark/pull/46960.

I guess so from the PR description.

File: python/pyspark/sql/dataframe.py [hunk 3]
```diff
@@ -5984,6 +5984,41 @@ def writeTo(self, table: str) -> DataFrameWriterV2:
...
+        Merges a set of updates, insertions, and deletions based on a source table into
+        a target table.
+
+        .. versionadded:: 4.0.0
+
+        Parameters
+        ----------
+        table : str
+            Target table name to merge into.
+        condition : :class:`Column`
```
Comment (viirya) [inline python/pyspark/sql/dataframe.py:5999]: This is `Column`, but the signature is `ColumnOrName`?

File: python/pyspark/sql/readwriter.py [hunk 4]
```diff
@@ -2465,6 +2465,200 @@ def overwritePartitions(self) -> None:
         self._jwriter.overwritePartitions()
 
 
+class MergeIntoWriter:
```
Comment (viirya) [inline python/pyspark/sql/readwriter.py:2468]: Maybe this could be in a separate file?

File: python/pyspark/sql/readwriter.py [hunk 5]
```diff
@@ -2465,6 +2465,200 @@ def overwritePartitions(self) -> None:
...
+        self, condition: Optional["ColumnOrName"] = None
+    ) -> "MergeIntoWriter.WhenNotMatchedBySource":
+        """
+        Initialize a `WhenNotMatchedBySource` action with a condition.
+        This `WhenNotMatchedBySource` action will be executed when a target row does not match any
+        rows in the source table based on the merge condition and the specified `condition`
+        is satisfied.
+        This `WhenNotMatchedBySource` can be followed by one of the following merge actions:
+          - `updateAll`: Update all the not matched target table rows with source dataset rows.
+          - `update(Map)`: Update all the not matched target table rows while changing only
```
Comment (huaxingao) [inline python/pyspark/sql/readwriter.py:2525]: nit: `update(Map)` -> `update(Dict)`

**Comment by huaxingao** [comment]
I quickly went through the PR, and it looks good overall. I echo @viirya's  suggestion to add a few more tests to make this more robust.

File: python/pyspark/sql/dataframe.py (hunk 3 above)
Comment (xupefei) [reply python/pyspark/sql/dataframe.py:5999]: This actually must be a `Column` under the hood -- when given a string it must be wrapped by an `expr()`. What's your suggestion here? Should we allow strings (and wrap them) to simplify the user experience, or ask the user to explicitly convert them to Columns?

File: python/pyspark/sql/tests/test_dataframe.py (hunk 1 above)
Comment (xupefei) [reply python/pyspark/sql/tests/test_dataframe.py:777]: Will do!

Comment (xupefei) [reply python/pyspark/sql/tests/test_dataframe.py:777]: Done!

File: python/pyspark/sql/dataframe.py (hunk 3 above)
Comment (xupefei) [reply python/pyspark/sql/dataframe.py:6001]: I changed the method to accept `Column`.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: python/pyspark/sql/dataframe.py (hunk 3 above)
Comment (viirya) [reply python/pyspark/sql/dataframe.py:6001]: Thank you.

**Review by viirya — COMMENTED** [review]
Looks good. Thanks @xupefei @HyukjinKwon @huaxingao @allisonwang-db

**Comment by yashgangrades** [comment]
Hi @HyukjinKwon  @viirya ,

It looks like this change has been merged, but I still don't see any documentation on mergeInto in the [Spark 3.1.3 Python API documentation](https://spark.apache.org/docs/3.1.3/api/python/reference/index.html).

**Comment by HyukjinKwon** [comment]
this was merged for Spark 4.0.0

**Comment by viirya** [comment]
New feature won't be backported to previous versions.

You should see the documentation after Spark 4.0.0 releases.

**Comment by yashgangrades** [comment]
@viirya Do you know when is the Spark 4.0.0 release scheduled?

**Comment by viirya** [comment]
I'm not sure. I think the original schedule is Jun, 2024. The 4.0.0 preview was released in Jun, 2024. But I don't find the date of the final release.

File: python/pyspark/sql/tests/test_dataframe.py [hunk 6]
```diff
@@ -742,6 +742,79 @@ def test_df_show(self):
             message_parameters={"arg_name": "truncate", "arg_type": "str"},
         )
 
+    def test_df_merge_into(self):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_dataframe.py:745]: This should actually be skipped if we build without test. let me make a quick followup

Comment (HyukjinKwon) [reply python/pyspark/sql/tests/test_dataframe.py:745]: https://github.com/apache/spark/pull/48404

**Pull request closed without merging**